File: java/src/main/replicant/Call.java
// Copyright 2009 Kiel Hodges
package replicant;

import java.util.*;

final class Call {

  private final Object                  mock;
  private final String                  methodName;
  private final List<? extends Object>  arguments;

  public Call(Object mock, String methodName, List<? extends Object> arguments) {
    this.mock = mock;
    this.methodName = methodName;
    this.arguments = arguments;
  }
  
  public String toString() {
    String argumentsString                = arguments.toString();
    String argumentsStringWithoutBrackets = argumentsString.substring(1, argumentsString.length() - 1);
    return mock + "." + methodName + '(' + argumentsStringWithoutBrackets + ')';
  }
  
  public boolean equals(Object object) {
    if (!(object instanceof Call))
      return false;
    
    Call that = (Call) object;
    return this.mock.equals(that.mock) && 
      this.methodName.equals(that.methodName) && 
      this.arguments.equals(that.arguments);
  }

  public int hashCode() {
    return methodName.hashCode();
  }

}

File: scala/src/main/scala/replicant/Call.scala
// Copyright 2011 Kiel Hodges
package replicant

sealed class Call private(val mock: Any, val methodName: String, val argLists: List[ArgList]) {
  
  def apply(args: Any) = new Call(mock, methodName, this.argLists :+ ArgList(args)) 
  
  override def equals(other: Any) = other match {
    case that: Call => this.mock == that.mock && this.methodName == that.methodName && this.argLists == that.argLists   
    case _ => false
  }
  
  override def hashCode = 41 * (41 * (41 + mock.hashCode) + methodName.hashCode) + argLists.hashCode 

  override def toString = mock + "." + methodName + argLists.mkString
  
}

object Call {

  def apply(mock: Any, methodName: String) = new Call(mock, methodName, Nil)

  def describe(calls: Traversable[Call]): String = 
    if (calls.isEmpty)
      "no calls"
    else
      calls.map("\n  " + _).mkString
}  
